File: Fruit Into Baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) 
    {
        int i = 0, j = 0, k = 2, sz = fruits.size(), maxm = 0 ;
        unordered_map<int, int> umap ;
        
        while(j<sz)
        {
            umap[fruits[j]]++ ;
            
            if(umap.size() <= k)
            {
                maxm = max(maxm, j-i+1) ; j++ ;
            }
            else if(umap.size() > k)
            {
                while(umap.size() > k)
                {
                    umap[fruits[i]]-- ;
                    if(umap[fruits[i]] == 0){ umap.erase(fruits[i]) ; }
                    i++ ;
                }
                j++ ;
            }
        }
        return maxm ;
    }
};

File: Fruit Into Baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int max = 0 ;
        HashMap<Integer,Integer> map = new HashMap<>();
        for(int start = 0 , end = 0 ; end<fruits.length ; ++end){
           map.put(fruits[end],map.getOrDefault(fruits[end],0)+1);
             if(map.size() <= 2){
                max = Math.max(max, end-start+1);
            }
            while(map.size()>2){
               map.put(fruits[start],map.get(fruits[start])-1);
                if(map.get(fruits[start])==0){
                    map.remove(fruits[start]);           
                }
                start++;
            } 
        }
        return max ;
   }
}
